### PR TITLE
New cache implementation

### PR DESCRIPTION
**Pull request by paulolieuthier**

### CODE REVIEW

**Review by racevedoo — APPROVED** [review]

File: src/save.ts [hunk 1]
```diff
@@ -1,13 +1,47 @@
...
+
+      core.info('Generating files list...');
+      const patterns = core
+        .getInput('path', { required: true })
+        .split('\n')
+        .map(s => s.trim())
+        .filter(x => x !== '');
+      const globber = await glob.create(patterns.join('\n'));
+      const files = await globber.glob();
+      await utils.writeToFile('CACHEFILE.txt', files.join('\n'));
```
Comment (t0rr3sp3dr0) [inline src/save.ts:36]: I think we should write this to file in a temporary path, like the ones generated by `mktemp`. Imagine the glob processes a rule `.`, tar would compress all the files in the current directory, including this one that you just created.

File: src/save.ts [hunk 2]
```diff
@@ -1,13 +1,47 @@
 import * as core from '@actions/core';
-
-import save from '../node_modules/cache/dist/save';
+import * as io from '@actions/io';
+import * as glob from '@actions/glob';
+import * as utils from './utils'
+import * as auth from './auth';
+import * as proc from './proc';
```
Comment (t0rr3sp3dr0) [inline src/save.ts:6]:
```suggestion
import * as glob from '@actions/glob';
import * as io from '@actions/io';

import * as auth from './auth';
import * as proc from './proc';
import * as utils from './utils';
```

File: src/restore.ts [hunk 3]
```diff
@@ -1,13 +1,33 @@
 import * as core from '@actions/core';
-
-import restore from '../node_modules/cache/dist/restore';
+import * as auth from './auth';
+import * as proc from './proc';
```
Comment (t0rr3sp3dr0) [inline src/restore.ts:3]:
```suggestion

import * as auth from './auth';
import * as proc from './proc';
```

File: src/auth.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import axios from 'axios';
+import * as utils from './utils';
```
Comment (t0rr3sp3dr0) [inline src/auth.ts:2]:
```suggestion
import axios from 'axios';

import * as utils from './utils';
```

File: src/auth.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  Upload = 'upload',
+}
+
+export class Response {
+  constructor(
+    public accessKeyId: string,
+    public secretAccessKey: string,
+    public sessionToken: string,
+    public s3ObjectPath: string,
+    public cacheHit: boolean
```
Comment (t0rr3sp3dr0) [inline src/auth.ts:26]:
```suggestion
    public cacheHit: boolean,
```

File: src/auth.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+export async function get(cacheKey: string, type: Type): Promise<Response> {
+  const url = `${BASE_URL}/${defaultCacheUrl}/assumeRole/${cacheKey}/${type}`;
+  const response = await axios.get(url, authReqOpts);
+
+  const data = new Response(
+    response.data['AccessKeyId'],
+    response.data['SecretAccessKey'],
+    response.data['SessionToken'],
+    response.data['ObjectS3URI'],
+    response.data['CacheHit']
```
Comment (t0rr3sp3dr0) [inline src/auth.ts:39]:
```suggestion
    response.data['CacheHit'],
```

File: src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    .toString('base64')
+    .replace(/\+/g, '-')
+    .replace(/\//g, '_')
+    .replace(/=/g, '');
+}
+
+export function writeToFile(path: string, content: string): Promise<void> {
+  return new Promise((resolve, reject) => {
+    fs.writeFile(path, content, error => {
+      if (error) reject(error);
```
Comment (t0rr3sp3dr0) [inline src/utils.ts:14]:
```suggestion
      if (error)
        reject(error);
```

**Review by t0rr3sp3dr0 — CHANGES_REQUESTED** [review]

File: src/save.ts (hunk 1 above)
Comment (paulolieuthier) [reply src/save.ts]: f595218

**Pull request merged**